libs/server/Custom/CustomCommandManagerSession.cs

```diff
@@ -13,53 +14,81 @@ internal sealed class CustomCommandManagerSession
         readonly CustomCommandManager customCommandManager;
 
         // These session specific arrays are indexed by the same ID as the arrays in CustomCommandManager
-        readonly (CustomTransactionProcedure, int)[] sessionTransactionProcMap;
-        readonly CustomProcedure[] sessionCustomProcMap;
-
+        ExpandableMap<CustomTransactionProcedureWithArity> sessionTransactionProcMap;
+        ExpandableMap<CustomProcedure> sessionCustomProcMap;
 
         public CustomCommandManagerSession(CustomCommandManager customCommandManager)
         {
             this.customCommandManager = customCommandManager;
-            sessionTransactionProcMap = new (CustomTransactionProcedure, int)[CustomCommandManager.MaxRegistrations];
-            sessionCustomProcMap = new CustomProcedure[CustomCommandManager.MaxRegistrations];
+            sessionTransactionProcMap = new ExpandableMap<CustomTransactionProcedureWithArity>(CustomCommandManager.MinMapSize, 0, byte.MaxValue);
```

Comment:
Why does CustomCommandManagerSession need to pay the cost of locking the ExpandableMap - the idea of creating a per session copy was to make the per-session critical path blazingly fast and lock free.

Reply:
Ah I see what you're saying, makes sense.